[ADD] cms_delete_content

simahawk opened the pull request:
Generic controllers and view for handling deletetion of contents. See REAME.

NOTE: depends on #10 (ready for merge)

simahawk committed:
[add] cms_delete_content

simahawk commented:
@Yajo @antespi @lasley @PCatinean your feedback is much appreciated here :)

yajo commented:
Please remove the #5 commits.

Also, on the implementation side, I'm not sure it's the best...

You could simply add a qweb template that adds a button to delete. A JS animation makes the button to trigger a XMLRPC call that deletes the current content. A submodule only would need to add this to any template:

```
<t t-call="cms_delete_content.button">
  <t t-records="records_to_delete"/>
  <!-- Optionally: -->
  <t t-set="confirm">
    <p>This is my custom confirmation message</p>
  </t>
</t>
```

You wouldn't need to create any models, controllers or mixins.

simahawk commented:
@Yajo you don't need to create anything in any case. You just need a delete button as reported in README
```
<a class="btn btn-danger" t-att-href="object.cms_delete_confirm_url">Delete</a>
```
That's it. OTOH we could display this in a modal instead of redirecting to delete confirm view.
This is needed ATM but I plan to add a default toolbar with such actions (edit/delete/publish/etc) and you won't need anything else ;)

yajo commented:
But what about if you want to customize the confirmation message? The readme states very complex instructions. Besides, with my suggestion, you'd remove lots of code because you would need no controllers nor models.

simahawk commented:
@Yajo I cleaned it up and added JS to handle delete confirmation in place w/ a modal.
The [README](https://github.com/simahawk/website-cms/blob/add-cms_delete_content/cms_delete_content/README.md) should be clear enough.

A note regarding having controllers and some code in models: makes customization easier. For instance when showing a delete message you can easily include current object details (name or other values).

guewen reviewed (approved):
(code review)

lasley commented:
@simahawk - Any thoughts on tests?

simahawk commented:
@lasley yeah... just a "thought" for now :stuck_out_tongue_closed_eyes: 
I will add some basic tests, sure ;)

yajo commented on cms_delete_content/models/website_published.py:
```diff
@@ -0,0 +1,46 @@
+    def _compute_cms_delete_url(self):
+        for item in self:
+            item.cms_delete_url = \
+                '/cms/delete/{}/{}'.format(item._name, item.id)
```
Use unicode string if calling `.format`

yajo commented on cms_delete_content/models/website_published.py:
```diff
@@ -0,0 +1,46 @@
+    def _compute_cms_delete_confirm_url(self):
+        for item in self:
+            item.cms_delete_confirm_url = \
+                '/cms/delete/{}/{}/confirm'.format(item._name, item.id)
```
Same here.

yajo commented on cms_delete_content/models/website_published.py:
```diff
@@ -0,0 +1,46 @@
+                '/cms/delete/{}/{}/confirm'.format(item._name, item.id)
+
+    # customize this per-model
+    cms_after_delete_url = '/'
```
This would be better a computed field too, to allow easy extensibility.

simahawk replied:
No issues AFAIK. I'm already using and overriding this and other attributes all around in my custom models/modules and is working fine. Any side effect that I'm not considering?

yajo commented on cms_delete_content/models/test_models.py:
```diff
@@ -0,0 +1,10 @@
+
+testing = tools.config.get('test_enable')
+
+if testing:
```
Can't you do this inside the test itself?

simahawk replied:
tried, does not work AFAIR :(
Unless you import test.test_models into ``module.__init__``. Which does not change that much I'd say.

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,58 @@
+    var _t = core._t;
+
+    $(document).ready(function () {
+      $(".cms_delete_confirm").click(function(evt){
```
This is not exactly "bad", but very not-odoo.

You should declare here an animation, which already does all this job for you, and is an Odoo Class (which includes extensibility and inheritability by default). Check [this example](https://github.com/OCA/website/blob/9.0/website_snippet_country_dropdown/static/src/js/website_snippet_country_dropdown.js#L10).

simahawk replied:
well, this is a 1st prototype and I could make it more extendable even without animation. Something like this https://github.com/simahawk/website-cms/blob/ae9819c1b64d86eaef4aa804749b811bd30a64be/cms_status_message/static/src/js/tool.js#L27
Which are the real benefits for using an animation (as long as this is not something you are going to edit w/ website builder)?

yajo replied:
Besides following the Odoo standard, which is a benefit by itself (you see one animation and you know how they all work), [you get all of this](https://github.com/odoo/odoo/blob/9.0/addons/web_editor/static/src/js/snippets.animation.js), among which there are:

- No need for `$(document).ready()`.
- You inherit/extend throught `web.Class`.
- `this.$`
- Auto-applied to all elements matching `this.selector`.

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,58 @@
+          var $modal = $(modal),
+              $form = $('form', $modal);
+          // handle modal form submit
+          $form.submit(function(evt){
```
Please avoid using too many lambdas. When you switch to animation format as said above, then better make use of proxied methods inside the class itself, so that they can be also inherited/extended (besides more maintainable). Check how it's done in [the above example again](https://github.com/OCA/website/blob/9.0/website_snippet_country_dropdown/static/src/js/website_snippet_country_dropdown.js#L20).

simahawk replied:
same above :)

yajo commented on cms_delete_content/controllers/main.py:
```diff
@@ -0,0 +1,77 @@
+
+class DeleteMixin(object):
+
+    delete_confirm_template = 'cms_delete_content.delete_confirm'
```
Odoo provides better inheritance/extension through methods than through attributes, so it will be better if you can change this for a method.

simahawk replied:
I use mixins exactly for this reason: it's kind of weird how inheritance works with controllers.

yajo replied:
OH didn't notice this was a mixin...

yajo reviewed (changes requested):
Much better pal, however JS could be much better using Odoo facilities.

simahawk commented:
@Yajo I refactored JS w/ animation See https://github.com/OCA/website-cms/pull/6/commits/d03239d75358e289190f0f2890b84e15c783da98. It works until form submit binding. I don't get why it doesn't hook the event. I tried several ways but none of the seemed to work.

There's no error. It simply does not register the event even if you attach it to the document.
If I bind it manually after the modal is showed I see it among event listeners in inspector.

The bind is applied as it was before: on the instantiated modal before appending it to body. 
Can you have a look at it?

simahawk commented:
@lasley can you check the comment above?

lasley commented:
@simahawk can you walk me through replication instructions? I'm unclear on the error.

simahawk commented:
@lasley it's really simple: ajax submit of form does not work. The binding doesn't happen. It used to work when JS was just a serie of promise/callback (till the 1 commit ago). Here's the diff:

BEFORE (worked)
![image](https://cloud.githubusercontent.com/assets/347149/24168953/5b88cd1c-0e7c-11e7-8e53-6dca597761a8.png)

AFTER/NOW (doesn't work)
![image](https://cloud.githubusercontent.com/assets/347149/24168982/73b459d8-0e7c-11e7-848d-2cb63846b3a1.png)

I don't see the event listeners in the inspector, and indeed is not called at all so you get redirected to the form action. 
If I manually hook the event after the modal is fully loaded via the inspector I get it working.

lasley commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -5,54 +5,65 @@ odoo.define('cms_delete_content.delete_confirm', function (require) {
+        });
+        return false;
+      },
+      handle_nodal_success: function(result){
```
s/handle_nodal_success/handle_modal_success

lasley commented:
Which handler isn't getting bound (click, submit, etc) & which element? Is there a way for me to replicate on Runbot?

simahawk commented:
@lasley nevermind: ajax calls were swallowing self/this variable scope. I fixed it and is now finally working w/ animations.... so even @Yajo should be happy now :stuck_out_tongue_closed_eyes:

yajo commented on cms_delete_content/README.md:
```diff
@@ -0,0 +1,113 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
```
You are mixing syntaxes. This is rst syntax, and this README is markdown.

pedrobaeza replied:
But this file should be RST if we want that Odoo will be able to handle it, so better to rename to README.rst.

simahawk replied:
@pedrobaeza are you sure?
https://github.com/OCA/OCB/blob/9.0/openerp/modules/module.py#L26
https://github.com/OCA/OCB/blob/10.0/odoo/modules/module.py#L28

Once we solve this I'm gonna fix the mixing ;)

pedrobaeza replied:
Yes, very sure. Odoo admits that file extensions, but actually parses all as RST. Better that they only admit *.rst extension IMO.

simahawk replied:
:man_facepalming: ok...

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,69 @@
+          method: "GET",
+          url: this.modal_url,
+          data: {}
+        }).done(function (modal) {
```
I'd appreciate if you either proxy this `done` or return the promise in the function.

simahawk replied:
can we consider all of these notes as "improvements"? We can delegate this stuff to a later PR, I think :)

yajo replied:
Yes, of course.

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,69 @@
+          data: data,
+          dataType: 'json',
+        }).done(function (result) {
+          self.handle_modal_success(result);
```
You can just `.done($.proxy(this.handle_modal_success, this))`.

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,69 @@
+        }).done(function (result) {
+          self.handle_modal_success(result);
+        });
+        return false;
```
Please don't. Return the promise.

yajo commented on cms_delete_content/static/src/js/delete_content.js:
```diff
@@ -0,0 +1,69 @@
+      },
+      handle_modal_success: function(result){
+        // hide modal
+        this.$modal.modal('hide');
```
I think it's better if you destroy it instead of just hiding it. It will never again be used, right?

yajo commented on cms_delete_content/models/website_published.py:
```diff
@@ -0,0 +1,46 @@
+                u'/cms/delete/{}/{}/confirm'.format(item._name, item.id)
+
+    # customize this per-model
+    cms_after_delete_url = '/'
```
I feel a method would fit better here. What if you want a different url depending on some specific condition?

simahawk replied:
This is supposed to be a default for a model. But, you can do conditional stuff in this way:
```
@property
def cms_after_delete_url(self):
    # do your stuff
```
or simpler: use the 'redirect' attribute for "on demand" changes. See https://github.com/simahawk/website-cms/blob/9619b81883f15b33d44582fefa84c38c9582a63a/cms_delete_content/controllers/main.py#L36.

yajo replied:
Are you sure these properties work as expected under Odoo metaclasses system? I feel you wouldn't have a `self.env` for instance when doing that...

simahawk replied:
yep, you have it.

yajo replied:
And what's the benefit on using this system instead of a method like everybody else in Odoo ecosystem, besides saving a `()` when calling it and saving 2 LOC? 🤔

simahawk replied:
is that a "religious" thing? :stuck_out_tongue: What's the benefit of "calling it" then? :thinking: 
Personally I found it useless in this case. To me, this sounds like a parameter, nothing more. 

In most of the cases you won't override it conditionally, since whenever you'll have the need - depending on user action - you can just append a `?redirect=`.

That said, I can move it to a method and make you sleep well tonite :stuck_out_tongue_closed_eyes:

simahawk replied:
Forgot to mention: the other "_url" attributes are fields, so you don't call them. Having it as an attribute sounds consistent w/ the other attributes. So, a computed field instead of a method would be better. It just sounded "overwhelming" in this case.

lasley replied:
Best of both worlds, `@property`?

yajo commented on cms_delete_content/README.md:
```diff
@@ -0,0 +1,113 @@
+
+.. image:: https://odoo-community.org/logo.png
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
```
Mixing syntaxes again.

yajo commented on cms_delete_content/README.md:
```diff
@@ -0,0 +1,113 @@
+# Bug Tracker
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/website-cms/issues>`_. In case of trouble, please
```
Mixing syntax

yajo commented on cms_delete_content/README.md:
```diff
@@ -0,0 +1,113 @@
+
+Note: if you want to customize it on demand for particular pages,
+or you are deleting an item from another page (like a management page)
+you can pass ``?redirect=`` in the url, like:
```
mixing syntax?

yajo commented on cms_delete_content/templates/delete_confirm.xml:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+
+    <template id="delete_confirm" name="CMS default delete confirm">
```
Wouldn't this template make more sense as a client-side template?

simahawk replied:
AFAIK a client-side template is only for the client while this can be used on both client and server side and I think it's easier to override it (as any other website template). In any way you'll have to load it in JS, so what's the gain here?

yajo replied:
JS templates are loaded on demand. Moving it to JS would mean that in your submodules you wouldn't need to call this template in your views. JS would detect the `.cms_delete_content` element, it would bind its events there, and it would load the template only when that element is found, render it only when needed, destroy it when no more needed.

simahawk replied:
Good to know :) Anyway, that template is already loaded on demand by the ajax call and I prefer to keep the logic in the controller (+mixin) rather than in the JS.

yajo reviewed (commented):
This seems more professional indeed 😉 

Some more comments to know your thoughts before approval 😊

simahawk commented:
@pedrobaeza readme fixed. Fixed status msg too OCA/website-cms/pull/12

simahawk commented:
tnx all :) Already squashed: can somebody hit the button when the build is finished?
